clang/runtime/dpct-rt/include/dpct/group_utils.hpp

```diff
@@ -527,6 +527,94 @@ class workgroup_load {
 private:
   uint8_t *_local_memory;
 };
+
+// Shuffles blocked arrangement of data across work items
+// template parameters :
+// ITEMS_PER_WORK_ITEM: size_t variable controlling the number of items per
+// thread/work_item
+// T:  datatype
+// Item : typename parameter resembling sycl::nd_item<3> .
+template <size_t ITEMS_PER_WORK_ITEM, typename T, typename Item>
+class workgroup_shuffle {
+public:
+  static size_t get_local_memory_size(size_t group_work_items) { return 0; }
+  workgroup_shuffle(uint8_t *local_memory) : _local_memory(local_memory) {}
+
+  T temp_storage[group_work_items];
+
+  enum shuffle_algorithm {
+
+    SHUFFLE_UP,
+    SHUFFLE_DOWN
+
+  };
+
+  // Shifts a blocked segment of workgroup items up by one item
+  __dpct_inline__ void shuffle_up(const Item &item,
+                                  T (&input)[ITEMS_PER_WORK_ITEM],
+                                  T (&prev)[ITEMS_PER_WORK_ITEM],
+                                  T &block_suffix) {
+
+    size_t linear_tid = item.get_local_linear_id();
+    size_t group_work_items = item.get_local_range().size();
+    temp_storage[linear_tid] = input[ITEMS_PER_WORK_ITEM - 1];
+
+    item.barrier(sycl::access::fence_space::local_space);
```

Comment:
The `nd_item::barrier` function does not exist in the SYCL2020 spec and will be removed sometime in the future. Instead, we should use `sycl::group_barrier` here.

The switch may also negatively impact performance as `sycl::group_barrier` does not accept an address space and may synchronize the global space as well.